torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1293,10 +1293,24 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # Average grad by world_size for consistency with PyTorch DDP.
                 param.grad.div_(self.gradient_predivide_factor)
 
+            grad = param.grad.data
             if param._is_sharded:  # type: ignore[attr-defined]
-                grad_flatten = torch.flatten(param.grad)
+                # We clear `param.grad` to permit repeated gradient
+                # computations when this FSDP module is called multiple times.
+                # This is to avoid a race among multiple re-entrant backward
```

Comment:
is there a unittest that can capture this race condition?

Reply:
After offline discussion, we decided to omit such a unit test for now. If there is a non-contrived re-entrant backward use case for FSDP, we can revisit.